### PR TITLE
Feat: Eth wallet contract implementation

### PR DESCRIPTION
**Pull request by birchmd**
This PR adds the eth wallet contract implementation which will be automatically deployed with eth implicit accounts. This is as per the design outlined in https://github.com/near/NEPs/issues/518 .

It is intentional that the wallet contract implementation is isolated as its own crate and not contained in the broader nearcore workspace. This allows it to be reviewed, tested and audited independently from the rest of the Near code. The crate includes integration tests written using `near-workspaces`. However, the nearcore integration test related to the wallet contract has also been updated to check that the contract is automatically deployed when creating an eth address and that it works as expected.

This will not be the last PR in this project because I am only adding the logic for the implementation. It still remains to setup the reproducible build pipeline and update the tests which check the contract hashes. Additionally, there are still a few details which need to be finalized: (1) the Ethereum chain ID that will be associated with Near and (2) the Near account ID that will have the eth address registrar contract deployed. The eth address registrar contract stores a reverse lookup of Ethereum-like address to Near account IDs. It is necessary for the wallet contract to detect faulty relayers.

However this (relatively large) PR can be reviewed and merged, and the points above will be addressed in much more manageable follow-up PRs.

### CODE REVIEW

File: Cargo.toml
Comment (nagisa) [inline Cargo.toml:124]: Is there a reason these are a git dependency?

We should avoid git dependencies where possible. The fact that this is referencing a tag suggests to me that these should be on crates.io.

Comment (birchmd) [reply Cargo.toml:124]: Aurora doesn't have a regular process for releasing crates to crates.io. So for now a git dep is required to use the crate. But this is easy to update in the future if Aurora does start releasing crates.

**Comment by staffik** [comment]
I started reviewing this but it might take a while. I plan to send feedback tomorrow.

File: Cargo.toml
Comment (nagisa) [reply Cargo.toml:124]: I think it should really be a not "if Aurora starts", but rather "integrating with the rest of the ecosystem ought to be a very strong incentive for Aurora to start releasing crates as they're expanding."

We can't really audit git tags, they are not reproducible and might just make the build process slower every time as cargo can't cache git repositories globally.

Does it block this PR? Probably not. But waiting for Aurora to realize these problems on their own before an action is taken is also not great.

File: Cargo.lock
Comment (nagisa) [inline Cargo.lock:902]: Please investigate upgrading this dependency.

File: Cargo.lock
Comment (nagisa) [inline Cargo.lock:7073]: [Hmph.](https://github.com/paritytech/libsecp256k1/pull/101) I know we don't do a best job of keeping nearcore's dependencies up-to-date ourselves, but this was one merge button away…

**Review by nagisa — COMMENTED** [review]
Q for Adam: do we need to have the contract checked-in? Why aren't we building it on demand and e.g. verifying its reproducibility through “other” means?

For wallet tests lets see if we can (re-)use the infrastructure in `near-test-contracts`. At least that way all the relevant stuff stays in the same spot. Ideally these contracts would get built on the fly like `test-contract-rs` is, but its not a too big of a deal if doing so is really difficult.

---

I don't feel qualified to review the contract itself, and I don't think I have the capacity at the moment to become able. It looks like Adam is already on the hook for that anyhow, but please definitely talk with N1 EMs to prioritize it accordingly if that's not the case.

File: Cargo.lock
Comment (birchmd) [reply Cargo.lock:902]: This has been [done on the Aurora side](https://github.com/aurora-is-near/aurora-engine/pull/907). But it hasn't been released as a tag yet on GitHub. This will be resolved when we next upgrade the Aurora dependency.

**Comment by bowenwang1996** [comment]
@staffik any updates on the review?

File: integration-tests/src/tests/client/features/wallet_contract.rs
Comment (staffik) [inline integration-tests/src/tests/client/features/wallet_contract.rs:290]: Anyone can add access key to ETH-implicit account?

File: runtime/near-wallet-contract/implementation/wallet-contract/src/internal.rs
Comment (staffik) [inline runtime/near-wallet-contract/implementation/wallet-contract/src/internal.rs:42]: Where do we check that keccak hash of the public key dervide from signature matches the ETH-implicit address?
I do not see it in `validate_tx_relayer_data`, is it inside ` let tx: NormalizedEthTransaction = tx_kind.try_into()?;` somehow?

File: runtime/near-wallet-contract/implementation/wallet-contract/src/near_action.rs
Comment (staffik) [inline runtime/near-wallet-contract/implementation/wallet-contract/src/near_action.rs:10]: If that is ok I would only include only the actions we actually support by the Wallet Contract.
Recently we introduced a new action for nonrefundable-storage and I wonder how do we plan to maintain the list here.

File: runtime/near-wallet-contract/implementation/wallet-contract/src/lib.rs
Comment (staffik) [inline runtime/near-wallet-contract/implementation/wallet-contract/src/lib.rs:242]: How would wallet contract know which address registrar SC it is gonna use?
Would it automatically match with the `AddressRegistrar` implemented in the `address-registrar` crate somehow (i.e. both modules deployed as a package in `Aptos` or `Sui`)?

File: runtime/near-wallet-contract/implementation/wallet-contract/src/lib.rs
Comment (staffik) [inline runtime/near-wallet-contract/implementation/wallet-contract/src/lib.rs:172]: Why 5 here?

**Review by staffik — COMMENTED** [review]
Great work, looks impressive!
Apologies for delay, we are absorbed in statelessnet issues.
I did not check test code yet and will look into the issue of checking-in the contract.

File: runtime/near-wallet-contract/implementation/wallet-contract/src/internal.rs
Comment (birchmd) [reply runtime/near-wallet-contract/implementation/wallet-contract/src/internal.rs:49]: It is the convention for Ethereum transactions that if `to` is not present (equal to `None`) then the transaction is a new contract deployment.

File: runtime/near-wallet-contract/implementation/wallet-contract/src/internal.rs
Comment (birchmd) [reply runtime/near-wallet-contract/implementation/wallet-contract/src/internal.rs:275]: No matter what, it is always true that any action is signed by the owner of the address because `rlp_execute` starts by decoding the RLP-encoded Ethereum transaction and checking its signature. The purpose of this check is to force external relayers to pay for the Near all fees associated with a transaction, not just the gas. This is similar to how meta-transactions work for normal (not eth-implicit) Near accounts.

The reason to allow self-calls is to provide a pathway for relayers to make users pay for their own Near fees. The flow would be for the relayer to have the user sign a transaction adding a function call access key which the relayer controls to the eth-implicit account. Then the relayer can sign transactions with that access key to have the wallet contract call its own `rlp_execute` method (of course the payload to that method must still be an Ethereum transaction signed by the user's key; so the relayer cannot use this to make the account do something the user did not intend).

File: runtime/near-wallet-contract/implementation/wallet-contract/src/internal.rs
Comment (birchmd) [reply runtime/near-wallet-contract/implementation/wallet-contract/src/internal.rs:223]: It is the convention for Ethereum transactions that if to is not present (equal to None) then the transaction is a new contract deployment.

File: runtime/near-wallet-contract/implementation/wallet-contract/src/internal.rs
Comment (birchmd) [reply runtime/near-wallet-contract/implementation/wallet-contract/src/internal.rs:106]: Naming is hard 😅 
I was using "hash" as a verb in this name as opposed to a noun. My thinking was that this function is hashing its input to create an address. But I can see how the name could be confusing. Maybe I'll just call it `account_id_to_address`.

File: runtime/near-wallet-contract/implementation/wallet-contract/src/internal.rs
Comment (birchmd) [reply runtime/near-wallet-contract/implementation/wallet-contract/src/internal.rs:116]: In the non-test setup we use the Near host function, but in the tests we use the implementation present in a library.

File: runtime/near-wallet-contract/implementation/wallet-contract/src/internal.rs
Comment (birchmd) [reply runtime/near-wallet-contract/implementation/wallet-contract/src/internal.rs:55]: No, we only allow each transaction to encode a single Near action.

File: runtime/near-wallet-contract/implementation/wallet-contract/src/internal.rs
Comment (birchmd) [reply runtime/near-wallet-contract/implementation/wallet-contract/src/internal.rs:37]: Currently there is no way to delete eth-implicit accounts because the `DeleteAccount` action is disallowed (unsupported action error) by `rlp_execute` and it is also not allowed to add a `FullAccessKey` to the account to get around the wallet contract.

So for now this is not a concern. We'll need to revisit this question if we do allow deleting these accounts in the future.

File: runtime/near-wallet-contract/implementation/wallet-contract/src/internal.rs
Comment (birchmd) [reply runtime/near-wallet-contract/implementation/wallet-contract/src/internal.rs:42]: Ethereum transactions only have a signature, they do not directly include the `address` field. Inside the conversion to `NormalizedEthTransaction` the `address` field is set by using the `ecrecover` function. Then in `validate_tx_relayer_data` we enforce that this recovered address matched the current account id.

File: runtime/near-wallet-contract/implementation/wallet-contract/src/near_action.rs
Comment (birchmd) [reply runtime/near-wallet-contract/implementation/wallet-contract/src/near_action.rs:10]: My idea with including all the actions was to make it cleat what actions are explicitly forbidden (as opposed to actions which may have been forgotten about). It is a good point about the maintainability though. Maybe I'll only include the ones which are used and include a comment that it is intentional some are left out.

File: runtime/near-wallet-contract/implementation/wallet-contract/src/lib.rs
Comment (birchmd) [reply runtime/near-wallet-contract/implementation/wallet-contract/src/lib.rs:242]: The account id for the registrar contract will be hardcoded into the wallet contract. This account id will be chosen in advance and be a locked account (ie without any access keys) with the registrar contract deployed on both testnet and mainnet.

File: runtime/near-wallet-contract/implementation/wallet-contract/src/eth_emulation.rs
Comment (birchmd) [reply runtime/near-wallet-contract/implementation/wallet-contract/src/eth_emulation.rs:52]: `ft_balance_of` is a simple operation (just one storage read), so 5 Tgas should be enough. On the other hand `ft_transfer` involves two storage reads and two storage writes so it needs more gas to complete successfully. 10 Tgas should be enough. For example, `AxuVFRtJ6XKzyLA1eWxJoGg12g6Z663WHH61bvejaeSv` is a $AURORA `ft_transfer` receipt which used less than 4 Tgas (see on [nearblocks](https://nearblocks.io/txns/9TxB6Cq4K36QLigLYxsbkMyLT6DRLMXtMy8RBiyEiLWj?tab=execution)).

File: runtime/near-wallet-contract/implementation/wallet-contract/src/lib.rs
Comment (birchmd) [reply runtime/near-wallet-contract/implementation/wallet-contract/src/lib.rs:172]: The call to `lookup` in the registrar contract is very simple, just one storage read. So 5 Tgas should be enough for it.

File: runtime/near-wallet-contract/implementation/address-registrar/src/lib.rs
Comment (birchmd) [reply runtime/near-wallet-contract/implementation/address-registrar/src/lib.rs:65]: Doc comments added in https://github.com/near/nearcore/pull/10850/commits/447feefc0a7f6f13aa13d15c85918131dadf02c2

File: runtime/near-wallet-contract/implementation/wallet-contract/src/near_action.rs
Comment (birchmd) [reply runtime/near-wallet-contract/implementation/wallet-contract/src/near_action.rs:13]: Unused actions removed in https://github.com/near/nearcore/pull/10850/commits/9bb90c09357d822db77d86637bf8e2ce61a7dbbc

File: runtime/near-wallet-contract/implementation/wallet-contract/src/internal.rs
Comment (birchmd) [reply runtime/near-wallet-contract/implementation/wallet-contract/src/internal.rs:94]: Name changed in https://github.com/near/nearcore/pull/10850/commits/7ed4741cc369655fe33dd7ae540202609cffd065

File: runtime/near-wallet-contract/implementation/wallet-contract/src/internal.rs
Comment (birchmd) [reply runtime/near-wallet-contract/implementation/wallet-contract/src/internal.rs:255]: Test added in https://github.com/near/nearcore/pull/10850/commits/7ed4741cc369655fe33dd7ae540202609cffd065

File: runtime/near-wallet-contract/implementation/wallet-contract/src/eth_emulation.rs
Comment (birchmd) [reply runtime/near-wallet-contract/implementation/wallet-contract/src/eth_emulation.rs]: In production there will not be any suffix because eth-implicit accounts are top-level. But for testing (eg in `near-workspaces`) I assume that all wallet contracts are deployed as sub-accounts of the same namespace, hence they all share the same suffix.

Comment (staffik) [reply runtime/near-wallet-contract/implementation/wallet-contract/src/eth_emulation.rs]: Could we add a short comment that the suffix stuff is for testing purposes?

File: runtime/near-wallet-contract/implementation/wallet-contract/src/eth_emulation.rs
Comment (staffik) [reply runtime/near-wallet-contract/implementation/wallet-contract/src/eth_emulation.rs:48]: I wonder what would be consequences if it is not enough gas. But if you are confident that's not a problem then I believe you :)

File: runtime/near-wallet-contract/implementation/wallet-contract/src/internal.rs
Comment (staffik) [reply runtime/near-wallet-contract/implementation/wallet-contract/src/internal.rs:52]: I mean single action that is not a transfer action.

File: runtime/near-wallet-contract/implementation/wallet-contract/src/internal.rs
Comment (staffik) [reply runtime/near-wallet-contract/implementation/wallet-contract/src/internal.rs:94]: Aah I see 😅

File: integration-tests/src/tests/client/features/wallet_contract.rs
Comment (staffik) [reply integration-tests/src/tests/client/features/wallet_contract.rs:290]: Ahh, I see now. I mislooked that it signed by secret_key for ETH-implicit address.

**Comment by staffik** [comment]
> Thanks for the review @staffik ! I think I have addressed all your comments now.
> 
> > the issue of checking-in the contract
> 
> I started working on this [here](https://github.com/aurora-is-near/nearcore/commit/57dd5ba007ea459db25ff7146bb9a8a11dfbeb5a). I was planning on pushing the commit in another PR after this one is merged.

There are 3 comment left and I will quickly look at test-related code now.

File: runtime/near-wallet-contract/implementation/wallet-contract/src/eth_emulation.rs
Comment (birchmd) [reply runtime/near-wallet-contract/implementation/wallet-contract/src/eth_emulation.rs]: Done in https://github.com/near/nearcore/pull/10850/commits/ee16f2bf75aabea09ecad4ccf1a44d6950ffd26b

File: runtime/near-wallet-contract/implementation/wallet-contract/src/internal.rs
Comment (birchmd) [reply runtime/near-wallet-contract/implementation/wallet-contract/src/internal.rs:52]: No, in this case we ignore the payload entirely; it could contain some action or not and it does not matter. The reason is because on Ethereum you can include arbitrary bytes as the payload for transfers between externally owned accounts (EOAs) (ie user accounts) and this is used in some wallet implementations to allow messaging between addresses. To be compatible with this use-case, when we detect that the target account is another eth-implicit account then we assume the payload of the transaction is arbitrary bytes and we do not parse it at all. An indexer could could still parse the payload though to display it as a message in the user's UI.

If someone set the payload to be the encoding of a Near action this would have no effect because the payload is entirely ignored by the wallet contract.

File: runtime/near-wallet-contract/implementation/wallet-contract/src/internal.rs
Comment (birchmd) [reply runtime/near-wallet-contract/implementation/wallet-contract/src/internal.rs]: Changed to `TODO(eth-implicit)` in https://github.com/near/nearcore/pull/10850/commits/fc96b2e8f6584421a338f62d889553320cfe505f

**Comment by staffik** [comment]
> > Thanks for the review @staffik ! I think I have addressed all your comments now.
> > > the issue of checking-in the contract
> > 
> > 
> > I started working on this [here](https://github.com/aurora-is-near/nearcore/commit/57dd5ba007ea459db25ff7146bb9a8a11dfbeb5a). I was planning on pushing the commit in another PR after this one is merged.
> 
> There are 3 comments left and I will quickly look at test-related code now.

Sorry, I underestimated, will finish tomorrow morning.
We will need an approval from near code owner to be able to merge.

File: runtime/near-wallet-contract/implementation/wallet-contract/src/tests/utils/test_context.rs
Comment (staffik) [inline runtime/near-wallet-contract/implementation/wallet-contract/src/tests/utils/test_context.rs:147]: Why do we need restoring here? `git checkout` looks fragile here and may cause issues if later used in CI where git commands are restricted.

**Review by staffik — COMMENTED** [review]
Second part of review.

File: runtime/near-wallet-contract/implementation/wallet-contract/src/tests/emulation.rs
Comment (birchmd) [reply runtime/near-wallet-contract/implementation/wallet-contract/src/tests/emulation.rs:83]: Yes, these parameters are currently not validated by the wallet contract. We could validate the `gas_limit` against the `prepaid_gas` host function in the Near runtime, but unfortunately there is no way to get the gas price from inside the runtime to validate against the transaction's gas price.

This does give relayers room to lie to users about the cost of transactions, but I don't think it is a critical vulnerability because the relayers do not profit from misrepresenting the gas price.

File: runtime/near-wallet-contract/implementation/wallet-contract/src/tests/emulation.rs
Comment (birchmd) [reply runtime/near-wallet-contract/implementation/wallet-contract/src/tests/emulation.rs:116]: The `value` field is not yoctoNEAR, it is 10^6 * yoctoNEAR. The reason for this is because $NEAR has 24 decimal places, but base tokens in EVM chains have 18 decimals.

When constructing a transaction the `value` is set equal to `yoctoNEAR / 10^6` (integer division) and the `yoctoNEAR % 10^6` remainder is listed directly in the Near action that is being encoded by the transaction. So if less than 1 million yoctoNEAR is being used by the transaction then the value is set to zero.

File: runtime/near-wallet-contract/implementation/wallet-contract/src/tests/relayer.rs
Comment (birchmd) [reply runtime/near-wallet-contract/implementation/wallet-contract/src/tests/relayer.rs:162]: Why do you prefer that way? Inlined format args are preferred [according to clippy](https://rust-lang.github.io/rust-clippy/master/index.html#uninlined_format_args).

File: runtime/near-wallet-contract/implementation/wallet-contract/src/tests/utils/test_context.rs
Comment (birchmd) [reply runtime/near-wallet-contract/implementation/wallet-contract/src/tests/utils/test_context.rs:147]: In the `near-workspaces` test the registrar contract is deployed to a random account (using  `dev-deploy`), so it needs to be dynamically set when compiling the wallet contract. This is handled by reading the account id from a file at compile time. Therefore that file is modified during the tests, but it should not stay modified after the tests are finished. An alternative to `git checkout` could be to read the file before changing it and remember the old value to restore manually. I can make this change.

File: runtime/near-wallet-contract/implementation/wallet-contract/src/tests/caller_error.rs
Comment (birchmd) [reply runtime/near-wallet-contract/implementation/wallet-contract/src/tests/caller_error.rs:51]: No, it fails because `action.value() == 1` and the `attached_deposit == 0` on the Near transaction which calls `rlp_execute`.

File: runtime/near-wallet-contract/implementation/wallet-contract/src/tests/caller_error.rs
Comment (birchmd) [reply runtime/near-wallet-contract/implementation/wallet-contract/src/tests/caller_error.rs:71]: Yes, that's correct.

File: runtime/near-wallet-contract/implementation/wallet-contract/src/tests/relayer.rs
Comment (staffik) [reply runtime/near-wallet-contract/implementation/wallet-contract/src/tests/relayer.rs:162]: Aah, sorry. For some reason I assumed it does not work.

File: runtime/near-wallet-contract/implementation/wallet-contract/src/tests/caller_error.rs
Comment (birchmd) [reply runtime/near-wallet-contract/implementation/wallet-contract/src/tests/caller_error.rs:102]: No, `tx.value` does define how much $NEAR will be used in the promise that is created by the wallet contract. The yoctoNEAR in the Transfer action is set to 0 because the transfer amount is 1 whole $NEAR (10^24 yoctoNEAR), so there is no remainder to list in the action (see above comment about decimals in EVM vs NEAR). 

The reason the final wallet balance is a little higher than the initial balance is because Near has the 30% reward to smart contracts. In this case the external account called the wallet contract so a portion of the gas burnt was given as $NEAR to the wallet.

**Review by staffik — APPROVED** [review]
LGTM, amazing work!

File: runtime/near-wallet-contract/implementation/wallet-contract/src/tests/user_error.rs
Comment (birchmd) [reply runtime/near-wallet-contract/implementation/wallet-contract/src/tests/user_error.rs]: Fixed in https://github.com/near/nearcore/pull/10850/commits/24bd5747e318e1fbfed6664410bef34c46afe150

File: runtime/near-wallet-contract/implementation/wallet-contract/src/tests/utils/test_context.rs
Comment (birchmd) [reply runtime/near-wallet-contract/implementation/wallet-contract/src/tests/utils/test_context.rs:146]: Done in https://github.com/near/nearcore/pull/10850/commits/b900f78f03b35c7aa330871fe671090a30353b4c

**Review by akhi3030 — APPROVED** [review]
approving to unblock.

**Pull request merged**
